File: backend/dormammu/state/operator_sync.py
"""Root-index and operator-mirror synchronisation for .dev/ state.

This module owns the mechanics of keeping the root ``.dev/session.json`` and
``.dev/workflow_state.json`` files in sync with the active session's copies,
and of mirroring operator-facing Markdown files (DASHBOARD.md, PLAN.md,
TASKS.md, WORKFLOWS.md) between the root and the session directory.

It is intentionally free of bootstrap-generation and session-lifecycle concerns.
"""
from __future__ import annotations

import contextlib
import shutil
import sys
from pathlib import Path
from typing import TYPE_CHECKING, Any, Generator, Mapping, Sequence

try:
    import fcntl as _fcntl
    _HAS_FCNTL = True
except ImportError:
    _fcntl = None  # type: ignore[assignment]
    _HAS_FCNTL = False

from dormammu._utils import iso_now as _iso_now
from dormammu.state.models import ManagedWorktreeState
from dormammu.state.persistence import ensure_json_file, read_json, write_json
from dormammu.state.tasks import parse_tasks_document

if TYPE_CHECKING:
    from dormammu.config import AppConfig


ROOT_OPERATOR_MIRROR_FILENAMES = (
    "DASHBOARD.md",
    "PLAN.md",
    "TASKS.md",
    "WORKFLOWS.md",
)


def _runtime_skill_summary(payload: Mapping[str, Any] | None) -> dict[str, Any] | None:
    if not isinstance(payload, Mapping):
        return None
    latest = payload.get("latest")
    if not isinstance(latest, Mapping):
        return None
    summary = latest.get("summary")
    profile = latest.get("profile")
    if not isinstance(summary, Mapping) or not isinstance(profile, Mapping):
        return None
    result = {
        "active_role": payload.get("active_role"),
        "profile_name": profile.get("name"),
        "profile_source": profile.get("source"),
    }
    for key in (
        "selected_count",
        "visible_count",
        "hidden_count",
        "preloaded_count",
        "missing_preload_count",
        "shadowed_count",
        "custom_visible_count",
        "interesting_for_operator",
    ):
        result[key] = summary.get(key)
    return result


class OperatorSync:
    """Synchronise root-index JSON and operator-facing Markdown mirrors.

    Parameters
    ----------
    config:
        Runtime application config.
    base_dev_dir:
        The root ``.dev/`` directory (not a session subdirectory).
    """

    def __init__(self, config: AppConfig, base_dev_dir: Path) -> None:
        self.config = config
        self.base_dev_dir = base_dev_dir

    # ------------------------------------------------------------------
    # File lock
    # ------------------------------------------------------------------

    @contextlib.contextmanager
    def root_index_lock(self) -> Generator[None, None, None]:
        """Acquire an exclusive file lock on the root .dev/ index files.

        Prevents concurrent writes from multiple dormammu sessions racing on
        ``.dev/session.json`` and ``.dev/workflow_state.json``.  Falls back to
        a no-op context on platforms that lack ``fcntl`` (e.g. Windows).
        """
        self.base_dev_dir.mkdir(parents=True, exist_ok=True)
        lock_path = self.base_dev_dir / ".dev_lock"
        with lock_path.open("a", encoding="utf-8") as lock_file:
            if _HAS_FCNTL:
                _fcntl.flock(lock_file, _fcntl.LOCK_EX)
            try:
                yield
            finally:
                if _HAS_FCNTL:
                    _fcntl.flock(lock_file, _fcntl.LOCK_UN)

    # ------------------------------------------------------------------
    # Root index write
    # ------------------------------------------------------------------

    def write_root_index_for_session(
        self,
        *,
        session_dev_dir: Path,
        session_id: str,
        state_root: str,
        timestamp: str,
        list_sessions_fn: Any,
    ) -> None:
        """Atomically update the root index under a file lock."""
        with self.root_index_lock():
            self._write_root_index_locked(
                session_dev_dir=session_dev_dir,
                session_id=session_id,
                state_root=state_root,
                timestamp=timestamp,
                list_sessions_fn=list_sessions_fn,
            )

    def _write_root_index_locked(
        self,
        *,
        session_dev_dir: Path,
        session_id: str,
        state_root: str,
        timestamp: str,
        list_sessions_fn: Any,
    ) -> None:
        session_state = read_json(session_dev_dir / "session.json")
        workflow_state = read_json(session_dev_dir / "workflow_state.json")
        session_worktrees = ManagedWorktreeState.from_dict(session_state.get("worktrees"))
        workflow_worktrees = ManagedWorktreeState.from_dict(workflow_state.get("worktrees"))

        session_defaults: dict[str, Any] = {
            "active_session_id": session_id,
            "default_session_id": session_id,
            "selected_at": timestamp,
            "updated_at": timestamp,
            "state_schema_version": session_state.get("state_schema_version"),
            "current_session": {
                "session_id": session_id,
                "state_root": state_root,
                "session_path": f"{state_root}/session.json",
                "workflow_path": f"{state_root}/workflow_state.json",
                "dashboard_path": f"{state_root}/DASHBOARD.md",
                "plan_path": f"{state_root}/PLAN.md",
                "tasks_path": f"{state_root}/TASKS.md",
                "logs_dir": f"{state_root}/logs",
                "goal": session_state.get("bootstrap", {}).get("goal"),
                "updated_at": session_state.get("updated_at"),
                "active_phase": session_state.get("active_phase"),
                "active_roadmap_phase_ids": session_state.get("active_roadmap_phase_ids", []),
                "active_worktree_id": session_worktrees.active_worktree_id,
                "managed_worktree_count": session_worktrees.managed_count,
                "runtime_skills": _runtime_skill_summary(session_state.get("runtime_skills")),
            },
        }
        workflow_defaults: dict[str, Any] = {
            "version": workflow_state.get("version", 1),
            "state_schema_version": workflow_state.get("state_schema_version"),
            "updated_at": timestamp,
            "mode": workflow_state.get("mode", "supervised"),
            "active_session_id": session_id,
            "default_session_id": session_id,
            "source_of_truth": {
                "goal": workflow_state.get("source_of_truth", {}).get("goal", []),
                "machine_state": ".dev/workflow_state.json",
                "operator_state": [],
                "session_machine_state": f"{state_root}/workflow_state.json",
                "session_operator_state": [
                    f"{state_root}/DASHBOARD.md",
                    f"{state_root}/PLAN.md",
                    f"{state_root}/TASKS.md",
                ],
            },
            "session_index": {
                "active_session_id": session_id,
                "sessions_dir": ".dev/sessions",
            },
            "current_session": {
                "session_id": session_id,
                "state_root": state_root,
                "workflow_path": f"{state_root}/workflow_state.json",
                "session_path": f"{state_root}/session.json",
                "tasks_path": f"{state_root}/TASKS.md",
                "goal": workflow_state.get("bootstrap", {}).get("goal"),
                "updated_at": workflow_state.get("updated_at"),
                "active_worktree_id": workflow_worktrees.active_worktree_id,
                "managed_worktree_count": workflow_worktrees.managed_count,
                "runtime_skills": _runtime_skill_summary(workflow_state.get("runtime_skills")),
            },
            "sessions": list_sessions_fn(),
        }

        ensure_json_file(self.base_dev_dir / "session.json", session_defaults)
        root_session = read_json(self.base_dev_dir / "session.json")
        root_session["state_schema_version"] = session_defaults["state_schema_version"]
        root_session["active_session_id"] = session_id
        root_session["default_session_id"] = session_id
        root_session["selected_at"] = timestamp
        root_session["updated_at"] = timestamp
        root_session["current_session"] = session_defaults["current_session"]
        write_json(self.base_dev_dir / "session.json", root_session)

        ensure_json_file(self.base_dev_dir / "workflow_state.json", workflow_defaults)
        root_workflow = read_json(self.base_dev_dir / "workflow_state.json")
        root_workflow["state_schema_version"] = workflow_defaults["state_schema_version"]
        root_workflow["updated_at"] = timestamp
        root_workflow["active_session_id"] = session_id
        root_workflow["default_session_id"] = session_id
        root_workflow["source_of_truth"] = workflow_defaults["source_of_truth"]
        root_workflow["session_index"] = workflow_defaults["session_index"]
        root_workflow["current_session"] = workflow_defaults["current_session"]
        root_workflow["sessions"] = list_sessions_fn()
        for key in ("bootstrap", "intake", "workflow_policy", "roadmap"):
            value = workflow_state.get(key)
            if isinstance(value, Mapping):
                root_workflow[key] = dict(value)
            elif key in root_workflow:
                root_workflow.pop(key, None)
        write_json(self.base_dev_dir / "workflow_state.json", root_workflow)

        self.sync_root_operator_mirrors(
            session_dev_dir=session_dev_dir,
            active_session_id=session_id,
        )

    # ------------------------------------------------------------------
    # Mirror sync
    # ------------------------------------------------------------------

    def _read_active_session_id(self) -> str | None:
        """Read the active session ID from the root index file."""
        from dormammu.state.session_manager import SessionManager

        return SessionManager.current_session_id(self.base_dev_dir / "session.json")

    def sync_root_operator_mirrors(
        self,
        *,
        session_dev_dir: Path,
        active_session_id: str,
    ) -> None:
        """Copy operator Markdown files from session dir to root ``.dev/``."""
        if self._read_active_session_id() != active_session_id:
            return
        self.base_dev_dir.mkdir(parents=True, exist_ok=True)
        for filename in ROOT_OPERATOR_MIRROR_FILENAMES:
            source = session_dev_dir / filename
            target = self.base_dev_dir / filename
            if source.exists():
                shutil.copy2(source, target)
            elif target.exists():
                target.unlink()

    def sync_active_root_operator_mirrors_into_session(
        self,
        *,
        session_dev_dir: Path,
        active_session_id: str,
    ) -> None:
        """Copy newer root operator Markdown files into the active session dir."""
        if self._read_active_session_id() != active_session_id:
            return
        for filename in ROOT_OPERATOR_MIRROR_FILENAMES:
            root_path = self.base_dev_dir / filename
            session_path = session_dev_dir / filename
            if not root_path.exists():
                continue
            if not session_path.exists():
                shutil.copy2(root_path, session_path)
                continue
            root_stat = root_path.stat()
            session_stat = session_path.stat()
            if root_stat.st_mtime_ns <= session_stat.st_mtime_ns:
                continue
            root_text = root_path.read_text(encoding="utf-8")
            session_text = session_path.read_text(encoding="utf-8")
            if root_text != session_text:
                shutil.copy2(root_path, session_path)

    # ------------------------------------------------------------------
    # Operator task sync
    # ------------------------------------------------------------------

    def refresh_active_roadmap_phase_ids(
        self,
        *,
        session_path: Path,
        workflow_path: Path,
        roadmap_phase_ids: Sequence[str],
        timestamp: str,
    ) -> None:
        """Persist the active roadmap phase IDs into both state files."""
        normalized_phase_ids = list(roadmap_phase_ids)

        session_state = read_json(session_path)
        session_state["updated_at"] = timestamp
        session_state["active_roadmap_phase_ids"] = normalized_phase_ids
        session_loop = session_state.get("loop")
        if isinstance(session_loop, dict):
            request_payload = session_loop.get("request")
            if isinstance(request_payload, dict):
                request_payload["expected_roadmap_phase_id"] = (
                    normalized_phase_ids[0] if normalized_phase_ids else None
                )
        write_json(session_path, session_state)

        workflow_state = read_json(workflow_path)
        workflow_state["updated_at"] = timestamp
        workflow_state.setdefault("roadmap", {})
        workflow_state["roadmap"]["active_phase_ids"] = normalized_phase_ids
        workflow_loop = workflow_state.get("loop")
        if isinstance(workflow_loop, dict):
            request_payload = workflow_loop.get("request")
            if isinstance(request_payload, dict):
                request_payload["expected_roadmap_phase_id"] = (
                    normalized_phase_ids[0] if normalized_phase_ids else None
                )
        write_json(workflow_path, workflow_state)

    def sync_operator_state(
        self,
        *,
        session_path: Path,
        workflow_path: Path,
        operator_task_path: Path,
        timestamp: str,
        dev_dir: Path,
        display_state_path_fn: Any,
    ) -> None:
        """Parse operator task document and write task-sync state into JSON."""
        resolved_path = self.resolve_operator_sync_source(
            preferred_path=operator_task_path,
            dev_dir=dev_dir,
        )
        operator_text = (
            resolved_path.read_text(encoding="utf-8") if resolved_path.exists() else ""
        )
        operator_mtime = resolved_path.stat().st_mtime if resolved_path.exists() else None

        if operator_mtime is not None and session_path.exists():
            try:
                stored = read_json(session_path)
                stored_mtime = stored.get("operator_state_mtime")
                if stored_mtime is not None and abs(operator_mtime - float(stored_mtime)) > 1.0:
                    print(
                        f"[dormammu] Warning: {resolved_path.name} was modified externally "
                        f"(stored mtime={stored_mtime:.3f}, current={operator_mtime:.3f}). "
                        "Manual edits will be preserved by re-reading the file.",
                        file=sys.stderr,
                    )
            except Exception:
                pass

        parsed_tasks = parse_tasks_document(
            operator_text,
            source=display_state_path_fn(resolved_path),
        )
        task_sync = parsed_tasks.current_workflow.to_dict(synced_at=timestamp)

        session_state = read_json(session_path)
        session_state["updated_at"] = timestamp
        session_state["task_sync"] = task_sync
        if operator_mtime is not None:
            session_state["operator_state_mtime"] = operator_mtime
        write_json(session_path, session_state)

        workflow_state = read_json(workflow_path)
        workflow_state["updated_at"] = timestamp
        workflow_state.setdefault("operator_sync", {})
        workflow_state["operator_sync"]["tasks"] = task_sync
        write_json(workflow_path, workflow_state)

    @staticmethod
    def resolve_operator_sync_source(
        *,
        preferred_path: Path,
        dev_dir: Path,
    ) -> Path:
        """Choose the best operator checklist source for task-sync state.

        Prefers the checklist file with the fewest pending ``- [ ]`` items
        (most progress) to avoid the supervisor seeing stale data when the
        agent has been updating PLAN.md rather than TASKS.md.
        """
        candidates: list[Path] = []
        seen: set[Path] = set()
        for path in (preferred_path, dev_dir / "PLAN.md", dev_dir / "TASKS.md"):
            resolved = path.resolve()
            if resolved in seen or not path.exists():
                continue
            seen.add(resolved)
            candidates.append(path)

        if not candidates:
            return preferred_path
        if len(candidates) == 1:
            return candidates[0]

        def _pending_count(path: Path) -> int:
            try:
                return path.read_text(encoding="utf-8").count("- [ ] ")
            except OSError:
                return 999

        best = min(
            candidates,
            key=lambda p: (_pending_count(p), -p.stat().st_mtime_ns, p.name != "TASKS.md"),
        )
        return best if best.exists() else preferred_path

File: runtime/src/state/operatorSync.ts
import {
  copyFile,
  mkdir,
  readFile,
  stat,
  unlink,
  writeFile
} from "node:fs/promises";
import path from "node:path";

import { ensureJsonFile, readJson, writeJson, type JsonObject } from "./persistence.js";
import { SessionManager, managedWorktreeSummaryFromDict } from "./sessionManager.js";
import { operatorTaskSyncToDict, parseTasksDocument } from "./tasks.js";

export const ROOT_OPERATOR_MIRROR_FILENAMES = [
  "DASHBOARD.md",
  "PLAN.md",
  "TASKS.md",
  "WORKFLOWS.md"
] as const;

type RuntimeSkillSummary = {
  active_role: unknown;
  profile_name: unknown;
  profile_source: unknown;
  selected_count: unknown;
  visible_count: unknown;
  hidden_count: unknown;
  preloaded_count: unknown;
  missing_preload_count: unknown;
  shadowed_count: unknown;
  custom_visible_count: unknown;
  interesting_for_operator: unknown;
};

async function exists(filePath: string): Promise<boolean> {
  try {
    await stat(filePath);
    return true;
  } catch (error) {
    if (
      error instanceof Error &&
      "code" in error &&
      (error as NodeJS.ErrnoException).code === "ENOENT"
    ) {
      return false;
    }
    throw error;
  }
}

function isRecord(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function getRecord(payload: JsonObject, key: string): JsonObject {
  const value = payload[key];
  return isRecord(value) ? value : {};
}

export function runtimeSkillSummary(payload: unknown): RuntimeSkillSummary | null {
  if (!isRecord(payload)) {
    return null;
  }
  const latest = payload.latest;
  if (!isRecord(latest)) {
    return null;
  }
  const summary = latest.summary;
  const profile = latest.profile;
  if (!isRecord(summary) || !isRecord(profile)) {
    return null;
  }
  return {
    active_role: payload.active_role,
    profile_name: profile.name,
    profile_source: profile.source,
    selected_count: summary.selected_count,
    visible_count: summary.visible_count,
    hidden_count: summary.hidden_count,
    preloaded_count: summary.preloaded_count,
    missing_preload_count: summary.missing_preload_count,
    shadowed_count: summary.shadowed_count,
    custom_visible_count: summary.custom_visible_count,
    interesting_for_operator: summary.interesting_for_operator
  };
}

async function copyIfPresent(source: string, target: string): Promise<void> {
  if (await exists(source)) {
    await copyFile(source, target);
  } else if (await exists(target)) {
    await unlink(target);
  }
}

async function mtimeNs(filePath: string): Promise<bigint> {
  const stats = await stat(filePath, { bigint: true });
  return stats.mtimeNs;
}

function pendingCount(text: string): number {
  return text.split("- [ ] ").length - 1;
}

let rootIndexQueue: Promise<void> = Promise.resolve();

export class OperatorSync {
  readonly baseDevDir: string;

  constructor(options: { baseDevDir: string }) {
    this.baseDevDir = options.baseDevDir;
  }

  async rootIndexLock<T>(run: () => Promise<T>): Promise<T> {
    await mkdir(this.baseDevDir, { recursive: true });
    await writeFile(path.join(this.baseDevDir, ".dev_lock"), "", { flag: "a" });
    const ready = rootIndexQueue.catch(() => undefined);
    let release: () => void = () => undefined;
    const held = new Promise<void>((resolve) => {
      release = resolve;
    });
    rootIndexQueue = ready.then(() => held);
    await ready;
    try {
      return await run();
    } finally {
      release();
    }
  }

  async writeRootIndexForSession(options: {
    sessionDevDir: string;
    sessionId: string;
    stateRoot: string;
    timestamp: string;
    listSessions: () => Promise<unknown[]> | unknown[];
  }): Promise<void> {
    await this.rootIndexLock(() => this.writeRootIndexLocked(options));
  }

  async writeRootIndexLocked(options: {
    sessionDevDir: string;
    sessionId: string;
    stateRoot: string;
    timestamp: string;
    listSessions: () => Promise<unknown[]> | unknown[];
  }): Promise<void> {
    const sessionState = await readJson(path.join(options.sessionDevDir, "session.json"));
    const workflowState = await readJson(path.join(options.sessionDevDir, "workflow_state.json"));
    const sessionWorktrees = managedWorktreeSummaryFromDict(sessionState.worktrees);
    const workflowWorktrees = managedWorktreeSummaryFromDict(workflowState.worktrees);
    const sessionBootstrap = getRecord(sessionState, "bootstrap");
    const workflowBootstrap = getRecord(workflowState, "bootstrap");
    const workflowSourceOfTruth = getRecord(workflowState, "source_of_truth");

    const sessionDefaults: JsonObject = {
      active_session_id: options.sessionId,
      default_session_id: options.sessionId,
      selected_at: options.timestamp,
      updated_at: options.timestamp,
      state_schema_version: sessionState.state_schema_version,
      current_session: {
        session_id: options.sessionId,
        state_root: options.stateRoot,
        session_path: `${options.stateRoot}/session.json`,
        workflow_path: `${options.stateRoot}/workflow_state.json`,
        dashboard_path: `${options.stateRoot}/DASHBOARD.md`,
        plan_path: `${options.stateRoot}/PLAN.md`,
        tasks_path: `${options.stateRoot}/TASKS.md`,
        logs_dir: `${options.stateRoot}/logs`,
        goal: sessionBootstrap.goal,
        updated_at: sessionState.updated_at,
        active_phase: sessionState.active_phase,
        active_roadmap_phase_ids: sessionState.active_roadmap_phase_ids ?? [],
        active_worktree_id: sessionWorktrees.activeWorktreeId,
        managed_worktree_count: sessionWorktrees.managedCount,
        runtime_skills: runtimeSkillSummary(sessionState.runtime_skills)
      }
    };
    const workflowDefaults: JsonObject = {
      version: workflowState.version ?? 1,
      state_schema_version: workflowState.state_schema_version,
      updated_at: options.timestamp,
      mode: workflowState.mode ?? "supervised",
      active_session_id: options.sessionId,
      default_session_id: options.sessionId,
      source_of_truth: {
        goal: workflowSourceOfTruth.goal ?? [],
        machine_state: ".dev/workflow_state.json",
        operator_state: [],
        session_machine_state: `${options.stateRoot}/workflow_state.json`,
        session_operator_state: [
          `${options.stateRoot}/DASHBOARD.md`,
          `${options.stateRoot}/PLAN.md`,
          `${options.stateRoot}/TASKS.md`
        ]
      },
      session_index: {
        active_session_id: options.sessionId,
        sessions_dir: ".dev/sessions"
      },
      current_session: {
        session_id: options.sessionId,
        state_root: options.stateRoot,
        workflow_path: `${options.stateRoot}/workflow_state.json`,
        session_path: `${options.stateRoot}/session.json`,
        tasks_path: `${options.stateRoot}/TASKS.md`,
        goal: workflowBootstrap.goal,
        updated_at: workflowState.updated_at,
        active_worktree_id: workflowWorktrees.activeWorktreeId,
        managed_worktree_count: workflowWorktrees.managedCount,
        runtime_skills: runtimeSkillSummary(workflowState.runtime_skills)
      },
      sessions: await options.listSessions()
    };

    const rootSessionPath = path.join(this.baseDevDir, "session.json");
    await ensureJsonFile(rootSessionPath, sessionDefaults);
    const rootSession = await readJson(rootSessionPath);
    rootSession.state_schema_version = sessionDefaults.state_schema_version;
    rootSession.active_session_id = options.sessionId;
    rootSession.default_session_id = options.sessionId;
    rootSession.selected_at = options.timestamp;
    rootSession.updated_at = options.timestamp;
    rootSession.current_session = sessionDefaults.current_session;
    await writeJson(rootSessionPath, rootSession);

    const rootWorkflowPath = path.join(this.baseDevDir, "workflow_state.json");
    await ensureJsonFile(rootWorkflowPath, workflowDefaults);
    const rootWorkflow = await readJson(rootWorkflowPath);
    rootWorkflow.state_schema_version = workflowDefaults.state_schema_version;
    rootWorkflow.updated_at = options.timestamp;
    rootWorkflow.active_session_id = options.sessionId;
    rootWorkflow.default_session_id = options.sessionId;
    rootWorkflow.source_of_truth = workflowDefaults.source_of_truth;
    rootWorkflow.session_index = workflowDefaults.session_index;
    rootWorkflow.current_session = workflowDefaults.current_session;
    rootWorkflow.sessions = await options.listSessions();
    for (const key of ["bootstrap", "intake", "workflow_policy", "roadmap"]) {
      const value = workflowState[key];
      if (isRecord(value)) {
        rootWorkflow[key] = { ...value };
      } else {
        delete rootWorkflow[key];
      }
    }
    await writeJson(rootWorkflowPath, rootWorkflow);

    await this.syncRootOperatorMirrors({
      sessionDevDir: options.sessionDevDir,
      activeSessionId: options.sessionId
    });
  }

  async readActiveSessionId(): Promise<string | null> {
    return SessionManager.currentSessionId(path.join(this.baseDevDir, "session.json"));
  }

  async syncRootOperatorMirrors(options: {
    sessionDevDir: string;
    activeSessionId: string;
  }): Promise<void> {
    if ((await this.readActiveSessionId()) !== options.activeSessionId) {
      return;
    }
    await mkdir(this.baseDevDir, { recursive: true });
    for (const filename of ROOT_OPERATOR_MIRROR_FILENAMES) {
      await copyIfPresent(
        path.join(options.sessionDevDir, filename),
        path.join(this.baseDevDir, filename)
      );
    }
  }

  async syncActiveRootOperatorMirrorsIntoSession(options: {
    sessionDevDir: string;
    activeSessionId: string;
  }): Promise<void> {
    if ((await this.readActiveSessionId()) !== options.activeSessionId) {
      return;
    }
    for (const filename of ROOT_OPERATOR_MIRROR_FILENAMES) {
      const rootPath = path.join(this.baseDevDir, filename);
      const sessionPath = path.join(options.sessionDevDir, filename);
      if (!(await exists(rootPath))) {
        continue;
      }
      if (!(await exists(sessionPath))) {
        await copyFile(rootPath, sessionPath);
        continue;
      }
      if ((await mtimeNs(rootPath)) <= (await mtimeNs(sessionPath))) {
        continue;
      }
      const [rootText, sessionText] = await Promise.all([
        readFile(rootPath, "utf8"),
        readFile(sessionPath, "utf8")
      ]);
      if (rootText !== sessionText) {
        await copyFile(rootPath, sessionPath);
      }
    }
  }

  async refreshActiveRoadmapPhaseIds(options: {
    sessionPath: string;
    workflowPath: string;
    roadmapPhaseIds: readonly string[];
    timestamp: string;
  }): Promise<void> {
    const normalizedPhaseIds = [...options.roadmapPhaseIds];

    const sessionState = await readJson(options.sessionPath);
    sessionState.updated_at = options.timestamp;
    sessionState.active_roadmap_phase_ids = normalizedPhaseIds;
    const sessionLoop = sessionState.loop;
    if (isRecord(sessionLoop) && isRecord(sessionLoop.request)) {
      sessionLoop.request.expected_roadmap_phase_id = normalizedPhaseIds[0] ?? null;
    }
    await writeJson(options.sessionPath, sessionState);

    const workflowState = await readJson(options.workflowPath);
    workflowState.updated_at = options.timestamp;
    const roadmap = isRecord(workflowState.roadmap) ? workflowState.roadmap : {};
    roadmap.active_phase_ids = normalizedPhaseIds;
    workflowState.roadmap = roadmap;
    const workflowLoop = workflowState.loop;
    if (isRecord(workflowLoop) && isRecord(workflowLoop.request)) {
      workflowLoop.request.expected_roadmap_phase_id = normalizedPhaseIds[0] ?? null;
    }
    await writeJson(options.workflowPath, workflowState);
  }

  async syncOperatorState(options: {
    sessionPath: string;
    workflowPath: string;
    operatorTaskPath: string;
    timestamp: string;
    devDir: string;
    displayStatePath: (filePath: string) => string;
    warn?: (message: string) => void;
  }): Promise<void> {
    const resolvedPath = await OperatorSync.resolveOperatorSyncSource({
      preferredPath: options.operatorTaskPath,
      devDir: options.devDir
    });
    const sourceExists = await exists(resolvedPath);
    const operatorText = sourceExists ? await readFile(resolvedPath, "utf8") : "";
    const operatorMtime = sourceExists ? (await stat(resolvedPath)).mtimeMs / 1000 : null;

    if (operatorMtime !== null && (await exists(options.sessionPath))) {
      try {
        const stored = await readJson(options.sessionPath);
        const storedMtime = stored.operator_state_mtime;
        if (
          storedMtime !== undefined &&
          Math.abs(operatorMtime - Number(storedMtime)) > 1.0
        ) {
          const message =
            `[dormammu] Warning: ${path.basename(resolvedPath)} was modified externally ` +
            `(stored mtime=${Number(storedMtime).toFixed(3)}, ` +
            `current=${operatorMtime.toFixed(3)}). ` +
            "Manual edits will be preserved by re-reading the file.";
          if (options.warn) {
            options.warn(message);
          } else {
            console.error(message);
          }
        }
      } catch {
        // Warning detection is best effort and must not block state sync.
      }
    }

    const parsedTasks = parseTasksDocument(operatorText, {
      source: options.displayStatePath(resolvedPath)
    });
    const taskSync = operatorTaskSyncToDict(parsedTasks.currentWorkflow, {
      syncedAt: options.timestamp
    });

    const sessionState = await readJson(options.sessionPath);
    sessionState.updated_at = options.timestamp;
    sessionState.task_sync = taskSync;
    if (operatorMtime !== null) {
      sessionState.operator_state_mtime = operatorMtime;
    }
    await writeJson(options.sessionPath, sessionState);

    const workflowState = await readJson(options.workflowPath);
    workflowState.updated_at = options.timestamp;
    const operatorSync = isRecord(workflowState.operator_sync)
      ? workflowState.operator_sync
      : {};
    operatorSync.tasks = taskSync;
    workflowState.operator_sync = operatorSync;
    await writeJson(options.workflowPath, workflowState);
  }

  static async resolveOperatorSyncSource(options: {
    preferredPath: string;
    devDir: string;
  }): Promise<string> {
    const candidates: string[] = [];
    const seen = new Set<string>();
    for (const candidate of [
      options.preferredPath,
      path.join(options.devDir, "PLAN.md"),
      path.join(options.devDir, "TASKS.md")
    ]) {
      const resolved = path.resolve(candidate);
      if (seen.has(resolved) || !(await exists(candidate))) {
        continue;
      }
      seen.add(resolved);
      candidates.push(candidate);
    }

    if (!candidates.length) {
      return options.preferredPath;
    }
    if (candidates.length === 1) {
      return candidates[0];
    }

    const ranked = await Promise.all(
      candidates.map(async (candidate) => {
        let pending = 999;
        try {
          pending = pendingCount(await readFile(candidate, "utf8"));
        } catch {
          pending = 999;
        }
        const stats = await stat(candidate, { bigint: true });
        return {
          candidate,
          pending,
          mtimeNs: stats.mtimeNs,
          tasksPenalty: path.basename(candidate) === "TASKS.md" ? 0 : 1
        };
      })
    );
    ranked.sort((left, right) => {
      if (left.pending !== right.pending) {
        return left.pending - right.pending;
      }
      if (left.mtimeNs !== right.mtimeNs) {
        return left.mtimeNs > right.mtimeNs ? -1 : 1;
      }
      return left.tasksPenalty - right.tasksPenalty;
    });
    return ranked[0].candidate;
  }
}
